jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/updates/TaskAssigningUpdates.java

```diff
@@ -0,0 +1,228 @@
+    private static final String ASSIGNED_TASKS_QUERY =
+            "select ti.taskId,ti.actualOwner from AuditTaskImpl ti where ti.actualOwner != '' and ti.status = 'Reserved'";
+    private static final String TASK_STARTED = "STARTED";
+    private static final String TASK_COMPLETED = "COMPLETED";
+    private static final long TASK_COMPLETION_DELAY_MILLIS = 5000L;
+
+    private final int processCount;
+    private final int taskCount;
+    private final int userCount;
+
+    private Set<Long> completedTasks;
+    private ExecutorService threadPoolExecutor;
+    private Timer taskAssignmentDuration;
+
+    public TaskAssigningUpdates(int processCount, int userCount) {
+        this.processCount = processCount;
+        this.taskCount = processCount * 3;
+        this.userCount = userCount;
+        if (taskCount < 2 * userCount) {
+            throw new IllegalArgumentException("For a relevance of the measured metric the number of tasks must be " +
+                    "at least 2 * number of users. Please note that the same number of users must be configured on " +
+                    "the KIE server before running this scenario");
+        }
+    }
+
+    private void beforeScenario() {
+        abortAllProcesses();
+        completedTasks = new ConcurrentHashMap<>().newKeySet();
+        threadPoolExecutor = Executors.newFixedThreadPool(userCount);
+    }
+
+    @Override
+    public void init() {
+        registerQuery(ASSIGNED_TASKS_QUERY_NAME, ASSIGNED_TASKS_QUERY);
+    }
+
+    @Override
+    public void initMetrics() {
+        MetricRegistry metrics = SharedMetricRegistry.getInstance();
+        taskAssignmentDuration = metrics.timer(MetricRegistry.name(getClass(), "taskassigning.update.task_assigned.duration"));
+    }
+
+    @Override
+    public void execute() {
+        beforeScenario();
+
+        // Create tasks by starting new processes.
+        startProcesses(processCount);
+
+        Set<Long> tasksInProgress = new ConcurrentHashMap<>().newKeySet();
+        // Keep completing tasks to introduce changes and trigger incremental re-planning.
+        while (completedTasks.size() < taskCount) {
+            TaskQueryFilterSpec filterSpec = new TaskQueryFilterSpec();
+            List<TaskInstance> assignedTasks = getQueryClient()
+                    .findHumanTasksWithFilters(ASSIGNED_TASKS_QUERY_NAME, filterSpec, 0, userCount);
+            if (assignedTasks.isEmpty()) { // If there are no tasks to work on, wait and retry.
+                sleep(500L);
+                continue;
+            }
+            LOGGER.debug(String.format("Completing %d tasks. Already finished %d tasks of %d.",
+                    assignedTasks.size(), completedTasks.size(), taskCount));
+            // Run each task in a separate thread.
+            for (TaskInstance task : assignedTasks) {
+                // Avoid submitting task some other thread already takes care of
+                if (!tasksInProgress.contains(task.getId())) {
```

Comment:
Maybe this is just a precaution but given that you iterate over a particular task just once, can it even happen?

Reply:
It's not just a precaution. Due to starting & completing tasks asynchronously, some tasks are not started before the next query in the outer loop on the line 120. As a result, such a task is retrieved again and once more submitted for start & complete, which fails as one of the threads working on the same task changes its state first.